### PR TITLE
BUG (string): Series.str.slice with negative step

### LINKED ISSUE #59710: API/BUG: obj.str.slice(start:None:-1)
```python
import pandas as pd
import pyarrow as pa

ser = pd.Series(["aafootwo", "aabartwo", pd.NA, "aabazqux"], dtype="string[pyarrow]")
ser2 = ser.astype(pd.ArrowDtype(pa.string()))

>>> ser.str.slice(None, None, -1)
0    owtoofaa
1    owtrabaa
2        <NA>
3    xuqzabaa
dtype: string

>>> ser2.str.slice(None, None, -1)
0       a
1       a
2    <NA>
3       a
dtype: string[pyarrow]
```

This looks to me like the ArrowEA version is wrong, can @mroeschke or @jorisvandenbossche confirm this is not intentional?

The ArrowEA version's only test is in `test_str_slice` in tests/extension/test_arrow.py, has no test cases with either stop=None or step<0.  The StringDtype version has a `test_slice` in tests/strings/test_strings.py that has a bit better coverage.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #59710 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -2419,6 +2419,10 @@ def _str_rpartition(self, sep: str, expand: bool) -> Self:
     def _str_slice(
         self, start: int | None = None, stop: int | None = None, step: int | None = None
     ) -> Self:
+        if step is not None and step < 0:
+            # GH#59710
+            result = self._apply_elementwise(lambda x: x[start:stop:step])
+            return type(self)(pa.chunked_array(result))
         if start is None:
             start = 0
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/arrow/array.py:2427]: See my comment in https://github.com/pandas-dev/pandas/issues/59710, but so probably this should be `start = -1` if `step < 0`, and then the above fallback might not be needed

**Comment by jorisvandenbossche** [comment]
If something fails for older version of pyarrow, can we instead fallback to the python implementation for that version of pyarrow? (instead of failing for the user, i.e. skipping that in the tests)

**Comment by jorisvandenbossche** [comment]
And while we are touching those methods, move it to the mixin at once as well?

**Comment by jbrockmendel** [comment]
> If something fails for older version of pyarrow, can we instead fallback to the python implementation for that version of pyarrow? (instead of failing for the user, i.e. skipping that in the tests)

just tried that and it broke the min versions test, looks like it is getting an int64 dtype somehow.  just restored the xfail.

**Comment by jorisvandenbossche** [comment]
Sorry, I don't understand how that can let those tests fail to use our long-used object dtype implementation. What was the error you got?  
FWIW it would help in reviewing if you pushed additional commits instead of reverting/force pushing, then it is clearer what you actually tried and what the CI said

**Comment by jorisvandenbossche** [comment]
> ive been rebasing (instead of merging main like usual)

Also when you add new commits for any change you make, you can still use rebase instead of merge to update with the latest main, if you prefer that

> on the theory it would make backporting easier.

PRs are always squashed into a single commit on top of main when merging, so I don't think this has any impact on backporting ease

File: pandas/tests/strings/test_strings.py [hunk 2]
```diff
@@ -393,7 +393,12 @@ def test_pipe_failures(any_string_dtype):
     [
         (2, 5, None, ["foo", "bar", np.nan, "baz"]),
         (0, 3, -1, ["", "", np.nan, ""]),
-        (None, None, -1, ["owtoofaa", "owtrabaa", np.nan, "xuqzabaa"]),
+        (
+            None,
+            None,
+            -1,
+            ["owtoofaa", "owtrabaa", np.nan, "xuqzabaa"],
+        ),
```
Comment (jorisvandenbossche) [inline pandas/tests/strings/test_strings.py:401]:
```suggestion
       (None, None, -1, ["owtoofaa", "owtrabaa", np.nan, "xuqzabaa"]),
```

File: doc/source/whatsnew/v2.3.0.rst [hunk 3]
```diff
@@ -103,8 +103,9 @@ Conversion
 Strings
 ^^^^^^^
 - Bug in :meth:`Series.str.replace` when ``n < 0`` for :class:`StringDtype` with ``storage="pyarrow"`` (:issue:`59628`)
+- Bug in ``ser.str.slice`` with negative ``step`` with :class:`ArrowDtype` giving incorrect results (:issue:`59710`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v2.3.0.rst:106]:
```suggestion
- Bug in ``ser.str.slice`` with negative ``step`` with :class:`ArrowDtype` and :class:`StringDtype` with ``storage="pyarrow"`` giving incorrect results (:issue:`59710`)
```

**Comment by jorisvandenbossche** [comment]
Just pushed one small additional case that wasn't covered yet (one that was failing before with StringDtype as well)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**